### ISSUE TITLE
problems about using gpu and seq2seq without attention

### ISSUE THREAD

**Issue by haichao592** [rsarxiv/seq2seq-attention#2]
When I running with command `th train.lua -cuda -attn -seq2seq`, I get the following error log:

> vocabulary size is 50000  
> source and target length is 25  
> number of parameters in the model: 55060000 
> cloning encoder 
> cloning embed  
> cloning decoder 
> cloning softmax 
> cloning criterion  
> totally needs training iterations 19564 
> /opt/torch/install/bin/luajit: /opt/torch/install/share/lua/5.1/nn/JoinTable.lua:24: 'for' limit must be a number
> stack traceback:
>     /opt/torch/install/share/lua/5.1/nn/JoinTable.lua:24: in function 'forward'
>     train.lua:147: in function 'opfunc'
>     /opt/torch/install/share/lua/5.1/optim/rmsprop.lua:32: in function 'rmsprop'
>     train.lua:231: in main chunk
>     [C]: in function 'dofile'
>     /opt/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
>     [C]: at 0x004064d0

When I running with command `th train.lua -cuda`, I get the following error log:

> vocabulary size is 50000  
> source and target length is 25  
> number of parameters in the model: 55310500 
> cloning encoder 
> cloning embed  
> cloning decoder 
> cloning softmax 
> cloning criterion  
> totally needs training iterations 19564 
> /opt/torch/install/bin/luajit: /opt/torch/install/share/lua/5.1/nn/MM.lua:42: invalid arguments: CudaTensor DoubleTensor CudaTensor 
> expected arguments: _CudaTensor_ CudaTensor~3D CudaTensor~3D
> stack traceback:
>     [C]: in function 'bmm'
>     /opt/torch/install/share/lua/5.1/nn/MM.lua:42: in function 'func'
>     /opt/torch/install/share/lua/5.1/nngraph/gmodule.lua:311: in function 'neteval'
>     /opt/torch/install/share/lua/5.1/nngraph/gmodule.lua:346: in function 'forward'
>     train.lua:157: in function 'opfunc'
>     /opt/torch/install/share/lua/5.1/optim/rmsprop.lua:32: in function 'rmsprop'
>     train.lua:231: in main chunk
>     [C]: in function 'dofile'
>     /opt/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
>     [C]: at 0x004064d0

**Comment by rsarxiv** [comment]
For problem one,  -seq2seq is the basic seq2seq model, and -attn means seq2seq+attention model, you needn't set them both, just use -seq2seq or -attn alone.

For problem two, maybe there are some bugs in the model code, I need to test them in the GPU environment, but I can't process it recently because I'm on a trip. I think you may try the basic seq2seq model via GPU.

Hope it helps.

**Comment by haichao592** [comment]
Have a good trip : )

For problem one,according to your code in train.lua
`cmd:option("-cuda",false,"")
cmd:option("-save_every",1000,"")
cmd:option("-seq2seq",false,"")
cmd:option("-attn",true,"")`
I want to run seq2seq without attention and if I remove seq2seq there will be 

> totally needs training iterations 19564  
> /opt/torch/install/bin/luajit: train.lua:136: attempt to index field 'encoder' (a nil value)
> stack traceback:
>     train.lua:136: in function 'opfunc'
>     /opt/torch/install/share/lua/5.1/optim/rmsprop.lua:32: in function 'rmsprop'
>     train.lua:231: in main chunk
>     [C]: in function 'dofile'
>     /opt/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk

**Comment by rsarxiv** [comment]
Hi, 
I've run my program just now, and it works well. So I think the problem is the preprocessing part, do you use the bot.py and the data I suppose or something else?

The attachment is the program running screenshot. 

发自 网易邮箱大师
On 07/03/2016 10:06, Haichao wrote:

For problem one,according to your code in train.lua
cmd:option("-cuda",false,"")
cmd:option("-save_every",1000,"")
cmd:option("-seq2seq",false,"")
cmd:option("-attn",true,"")
I want to run seq2seq without attention and if I remove seq2seq there will be

totally needs training iterations 19564

/opt/torch/install/bin/luajit: train.lua:136: attempt to index field 'encoder' (a nil value)
stack traceback:
train.lua:136: in function 'opfunc'
/opt/torch/install/share/lua/5.1/optim/rmsprop.lua:32: in function 'rmsprop'
train.lua:231: in main chunk
[C]: in function 'dofile'
/opt/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk

**Comment by rsarxiv** [comment]
@haichao592  I've fixed the GPU bugs just now, thank you for your problem. Hope it helps.
